enhancements/cvo/upgrade_preflight_checks.md

```diff
@@ -0,0 +1,169 @@
+---
+title: upgrade-preflight-checks-for-operators
+authors:
+  - "@LalatenduMohanty"
+reviewers:
+  - "@dgoodwin"
+  - "@crawford"
+  - "@sdodson"
+  - "@wking"
+approvers:
+  - "@crawford"
+  - "@sdodson"
+creation-date: 2020-06-04
+last-updated: 2020-09-02
+status: implementable
+---
+
+# Upgrade Preflight Checks For Operators
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [x] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+The goal of OpenShift 4 is to make upgrades seamless and risk free.
+To improve the upgrade success we should be able to detect if the new version is compatible with the existing version before running the actual upgrade.
+As OpenShift V4 is composed of many operators, we need to give the operators ability to run some checks before the upgrade to find out if the new incoming version is compatible with the existing version.
+
+## Motivation
+
+CVO managed operators should be able to run preflight checks before going through the actual upgrade.
+It would help operators to find out if they can move to the specific version safely before doing an actual upgrade.
+Also administrators are empowered check if upgrade to the target version is going to be successfully without doing an upgrade.
+Use preflight checks to improve the success rate of upgrades.
+
+
+### Goals
```

Comment:
In general I am -1 to generic frameworks for things that operators should encapsulate.  I do not want to see complex abstractions for executing code that could a) be encapsulated in the operator, b) are not going to be sufficiently common to do them frequently (thus not justifying a fully generic mechanism), or c) are no better than just encoding code in the CVO.

In fact, given all possible options, I would rather add O(10) specific code fragments to CVO in a z release over doing ANY generic framework.  The CVO was designed with the intent that if we needed specific workarounds, we could do them in CVO.  The CVO is not a general purpose component for tens of distributions, it is the core openshift operator.  While I don't want sprawl, I see no problem with specific time limited workarounds.

Reply:
While I generally agree with your sentiment, I think CVO is a bit of a special case. Due to it being a single point of failure for updates, it is critical that this code base remain as simple as possible. Otherwise, we run the risk of clusters becoming wedged and unable to apply or even check for further updates. By their nature, these pre-flight checks would be written by a number of developers, across teams, who are likely not as familiar with the CVO and the assumptions that have been made. Additionally, this coupling of operators to CVO through pre-flight checks feels like a step backward from the operator model we have today.